chain/network-primitives/src/types.rs

```diff
@@ -418,11 +419,13 @@ impl RoutedMessage {
         source: &PeerId,
         body: &RoutedMessageBody,
     ) -> CryptoHash {
-        hash(
-            &RoutedMessageNoSignature { target, author: source, body }
-                .try_to_vec()
-                .expect("Failed to serialize"),
```

Comment:
I feel like there might be a ton of places where we might want something like this, so let's extract this into function. 

```rust
pub fn hash(data: &[u8]) -> CryptoHash { }
pub fn hash_borsh<T: BorshSerialize>(value: &T) -> CryptoHash { }
```

In general, I feel that 99% of usages of `.try_to_vec` are suboptimal -- you almost never want to get a freshly allocated vec, you almost always want to serialized to some writer.

Reply:
Aside: I'd probably even argue for 

```
impl CryptoHash {
  pub fn hash_read<R: Read>(rdr: R) -> io::Result<CryptoHash> { }
  pub fn hash_bytes(bytes: &[u8]) -> CryptoHash {}
  pub fn hash_borsh<T: BorshSerialize>(value: &T) -> CryptoHash { }
}
```

API if designing the thing from scratch, but let's not change the cowpaths we already have.

Reply:
>   
> In general, I feel that 99% of usages of .try_to_vec are suboptimal -- you almost never want to get a freshly allocated vec, you almost always want to serialized to some writer.
> ```


Yes, that's why i'm doing this change. I'm removing it from `near-network`. We can ship a separate PRs to remove it from other places

>   pub fn hash_bytes(bytes: &[u8]) -> CryptoHash {}
>   pub fn hash_borsh<T: BorshSerialize>(value: &T) -> CryptoHash { }

I added those two above. I can can ```
  pub fn hash_read<R: Read>(rdr: R) -> io::Result<CryptoHash> { }
```
as well, but I don't see any usages of it at the moment.